pkg/grpc/api.go

```diff
@@ -87,38 +85,20 @@ type api struct {
 
 // APIOpts contains options for NewAPI.
 type APIOpts struct {
-	AppID                       string
-	AppChannel                  channel.AppChannel
-	Resiliency                  resiliency.Provider
-	CompStore                   *compstore.ComponentStore
-	PubsubAdapter               runtimePubsub.Adapter
-	DirectMessaging             messaging.DirectMessaging
-	Actors                      actors.Actors
-	SendToOutputBindingFn       func(name string, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error)
-	TracingSpec                 config.TracingSpec
-	AccessControlList           *config.AccessControlList
-	Shutdown                    func()
-	GetComponentsCapabilitiesFn func() map[string][]string
-	AppConnectionConfig         config.AppConnectionConfig
-	GlobalConfig                *config.Configuration
+	UniversalAPI          *universalapi.UniversalAPI
```

Comment:
Please revert so that these options are passed by value, and _not_ a pointer to an instantiated `*universalapi.UniversalAPI`. I don't think the code needs to be changed here at all, and is going against the recent efforts to decouple packages.

Ideally, `universalapi.UniversalAPI` itself should not have any exported fields, take in an `Options`, and use `funcs` to return fields needed by consumers.

This is part of the larger effort to ensure there is proper separation of concerns between packages WRT memory manipulation,

Reply:
I'm afraid I cannot do what you're asking.

`universalapi.UniversalAPI` must have exported fields because `grpc.api` inherits from that, and it's in a separate package.

Having the same struct (a pointer) is needed to make it so things like metadata API access the same metadata properties.

Reply:
> universalapi.UniversalAPI must have exported fields because grpc.api inherits from that, and it's in a separate package.

Reply:
Can we do that in a separate PR? That seems to be a rather large change

Reply:
I had a go at doing this [here](https://github.com/ItalyPaleAle/dapr/pull/97), PTAL

Reply:
Although I do like the changes, please let’s do that in a separate PR in Dapr/dapr so this bug fix can be merged more quickly.